Fix/failsafe dependency upgrade fix test case

debjit-hyland opened the pull request:
Testing PR builds to fix previous failure issues in - https://github.com/nuxeo/nuxeo-ai/pull/637

sainadha-krishna committed:
Updating to use latest platform HF 2023.16

sainadha-krishna committed:
Fixing test case failures - I

debjit-hyland committed:
Testing PR builds to analyse and fix test failures

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/nuxeo/nuxeo-ai?pullRequest=639) <br/>All committers have signed the CLA.

debjit-hyland committed:
Updating build workflow aligned to Node@20

debjit-hyland committed:
Updating build workflow maven run command to test concurrency issue

debjit-hyland committed:
NXCON-2: Fixed broken test cases

troger commented on pom.xml:
```diff
@@ -4,7 +4,7 @@
     <groupId>org.nuxeo</groupId>
     <artifactId>nuxeo-parent</artifactId>
-    <version>2023.0</version>
+    <version>2023.16</version>
```
While we're at upgrading, I think you could align here on the latest 2023 version, `2023.30`

debjit-hyland replied:
done

debjit-hyland committed:
Upgrading to use latest platform HF 2023.30

debjit-hyland committed:
Upgrading grpc-netty version to latest stable 1.64.0 compatible with protobuf-java version 3.25.7

debjit-hyland committed:
Removing deprecated method after protobuf upgrade and build command changes to make build faster

debjit-hyland committed:
Reverting Build with Maven command changes due to locking issue

debjit-hyland committed:
Generated fresh java files from proto files with protobuf 4.29.3 compiler to remove deprecated and vulnerable resources.

debjit-hyland committed:
-Include .proto files in project directory
-Remove commented code snippets
-add readme.md file for .proto to .java files generation process

debjit-hyland commented:
Things I want to achieve through this PR - 

- Bump Nuxeo LTS version to LTS 2023 latest (2023.31)
- Bump build workflow to run on Node@20
- Remove org.tensorflow.proto dependency from ai-core pom.xml module as it is deprecated and having java classes generated from old 3.5.1 protobuf.
- Include com.google.protobuf latest version in parent pom.xml
- Use .proto files from tensorflow to generate latest java files using protoc 29.3. This mitigates DOS vulnerability which was reported in the build warnings - https://github.com/protocolbuffers/protobuf/security/advisories/GHSA-h4h5-3hr4-j3g2
- Add readme.md file with steps to generate proto files in future using latest protoc.

FYI - org.tensorflow.proto dependency was removed as it is not maintained anymore and the last version(1.15.0) had vulnerabilities reported due to usage of old .java classes generated internally using protobuf - 3.5.1 (See -  https://github.com/protocolbuffers/protobuf/security/advisories/GHSA-h4h5-3hr4-j3g2). 

I have included the proto files in the codebase which is used to generate the latest java files using latest protoc. This approach is chosen keeping in mind the usecase of the Protobuf message types used in the codebase remains as is.

debjit-hyland commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/transcribe/AudioTranscription.java:
This will be removed on final commit before merge. Pls ignore as of now _/\_

kevinleturc commented on nuxeo-ai-pipes/src/main/java/org/nuxeo/ai/pipes/functions/PropertyUtils.java:
```diff
@@ -126,7 +126,7 @@ public static <T> T getPropertyValue(DocumentModel doc, String propertyName, Cla
     public static String base64EncodeBlob(Blob blob) {
         try {
-            if (blob != null) {
+            if (blob != null && blob.getByteArray() != null && blob.getByteArray().length > 0) {
                 return Base64.encodeBase64String(blob.getByteArray());
             }
```
An `InputStream` is created/fetched for each call to `getByArray`, this may be not performant to call it multiple times (depending on the blob implementation).
You should store the result:
```suggestion
            if (blob != null) {
                byte[] blobBytes = blob.getByteArray();
                 if (ArrayUtils.isNotEmpty(blobBytes)) {
                    return Base64.encodeBase64String(blobBytes);
                }
             }
```

kevinleturc commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichingStreamProcessor.java:
```diff
@@ -196,20 +202,35 @@ protected Collection<? extends AIMetadata> cacheGet(String cacheKey) {
+
+            });
+
+            FailsafeExecutor<Collection<AIMetadata>> executor = Failsafe.with(retryPolicy, circuitBreaker, timeout)
+                                                                    .onSuccess(r -> {
+                                                                        metrics.success();
+                                                                        if (log.isDebugEnabled()) {
+                                                                            log.debug("Enrichment result is " + r);
+                                                                        }
+                                                                    })
+                                                                    .onFailure(failure -> {
+                                                                        //this is final failure after all retries
+                                                                        metrics.error();
+                                                                        log.warn("Enrichment error ({}) for record: {}", enricherName, record, failure);
+                                                                    });
+
+            return executor.get(supplier);
```
```suggestion
            return Failsafe.with(retryPolicy, circuitBreaker, timeout)
                                                                    .onSuccess(r -> {
                                                                        metrics.success();
                                                                        if (log.isDebugEnabled()) {
                                                                            log.debug("Enrichment result is " + r);
                                                                        }
                                                                    })
                                                                    .onFailure(failure -> {
                                                                        //this is final failure after all retries
                                                                        metrics.error();
                                                                        log.warn("Enrichment error ({}) for record: {}", enricherName, record, failure);
                                                                    })
                                                                    .get(supplier);
```

debjit-hyland replied:
Reverted the inline Failsafe.with(...).get(supplier) change due to a type inference issue when using multiple policies. The generic type <Collection<AIMetadata>> causes a compilation error (Wrong number of type arguments: 1; required: 2) because the overloaded Failsafe.with(...) method expects two type parameters in this context.

Keeping the original version with an explicitly typed FailsafeExecutor<Collection<AIMetadata>> avoids ambiguity and ensures type safety without requiring additional casting or verbose generics.

kevinleturc replied:
Closing this comment and opening a new one on the same section.

ataillefer reviewed (commented):
Sounds good globally. Two points though:
- I assume that the commented code and comments about testing will be removed in the end?
- About the generated code, could it be genereated during the build pipeline and put inside a dedicated folder (e.g.: `target/generated-sources`), ignored by Git? Maybe using the Maven `generate-sources`phase.

guirenard commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichingStreamProcessor.java:
```diff
@@ -196,20 +202,35 @@ protected Collection<? extends AIMetadata> cacheGet(String cacheKey) {
+                                                                        metrics.success();
+                                                                        if (log.isDebugEnabled()) {
+                                                                            log.debug("Enrichment result is " + r);
+                                                                        }
```
```suggestion
                                                                        log.debug("Enrichment result is {}", r);
```

guirenard commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/tensorflow/TFRecord.java:
```diff
@@ -17,30 +17,25 @@
  * Partial implementation of a Protobuf object that represents Tensorflow record
  */
-public class TFRecord extends GeneratedMessageV3 implements MessageOrBuilder {
+
```
a bunch of emtpy lines in this class, we don't have empty lines between method and javadoc

guirenard commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/tensorflow/TFRecord.java:
```diff
@@ -73,11 +68,13 @@ public TFRecord(String docId, Features features) {
 
-    /**
+
+/**
```
bad indent

guirenard commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/tensorflow/TFRecord.java:
```diff
@@ -176,13 +173,15 @@ public int getSerializedSize() {
 
-    /**
+
+/**
```
bad indent

debjit-hyland committed:
Automating the java file generations from proto files during build.

debjit-hyland committed:
Removed commented codes, wrong indent and unnecesary spaces.

debjit-hyland committed:
- Included code suggestions in review
- Updated the readme.md for proto with updated details
- Removed commented code

debjit-hyland committed:
Reverting suggestion

kevinleturc commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichmentProvider.java:
```diff
@@ -56,19 +61,19 @@ public interface EnrichmentProvider {
      */
     @SuppressWarnings("unchecked")
     default RetryPolicy getRetryPolicy() {
```
Types should now be given to `RetryPolicy`:
```suggestion
    default RetryPolicy<Collection<AIMetadata>> getRetryPolicy() {
```
Ideally all implementation should receive type too.

kevinleturc commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichmentProvider.java:
```diff
@@ -56,19 +61,19 @@ public interface EnrichmentProvider {
     /**
      * The circuit breaker for the provider
      */
     default CircuitBreaker getCircuitBreaker() {
```
Same thing than getRetryPolicy:
```suggestion
    default CircuitBreaker<Collection<AIMetadata>> getCircuitBreaker() {
```

kevinleturc commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichingStreamProcessor.java:
`retryPolicy` and `circuitBreaker` attributes should receive their types:
```
        protected RetryPolicy<Collection<AIMetadata>> retryPolicy;

        protected CircuitBreaker<Collection<AIMetadata>> circuitBreaker;
```

kevinleturc commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichingStreamProcessor.java:
```diff
@@ -196,20 +202,31 @@ protected Collection<? extends AIMetadata> cacheGet(String cacheKey) {
+            CheckedSupplier<Collection<AIMetadata>> supplier = callable::call;
+            Timeout<Collection<AIMetadata>> timeout = Timeout.of(Duration.ofSeconds(60));
+
+            retryPolicy.onRetry(c -> {
```
I'm not sure this is what we want, as I understand it the given listener will be registered to the `retryPolicy` and not the ongoing failsafe execution, the retry policy should be copied.

See next comment.

kevinleturc commented on nuxeo-ai-core/src/main/java/org/nuxeo/ai/enrichment/EnrichingStreamProcessor.java:
```diff
@@ -196,20 +202,31 @@ protected Collection<? extends AIMetadata> cacheGet(String cacheKey) {
+
+            });
+
+            FailsafeExecutor<Collection<AIMetadata>> executor = Failsafe.with(retryPolicy, circuitBreaker, timeout)
+                    .onSuccess(r -> {
+                        metrics.success();
+                        log.debug("Enrichment result is: {}", r);
+                    })
+                    .onFailure(failure -> {
+                        //this is final failure after all retries
+                        metrics.error();
+                        log.warn("Enrichment error ({}) for record: {}", enricherName, record, failure);
+                    });
+
+            return executor.get(supplier);
```
When types are given to methods, attributes and others, the whole method could be re-wrote as:
```
        protected Collection<AIMetadata> callProvider(Record record, Callable<Collection<AIMetadata>> callable) {

            Timeout<Collection<AIMetadata>> timeout = Timeout.of(Duration.ofSeconds(60));

            var retryPolicy = this.retryPolicy.copy().onRetry(c -> {
                metrics.retry();
                log.debug("Retrying record: {}", record);
                //If we are retrying it means it has errorred out so metric has to be taken care of in higher versions of failsafe
                metrics.error();
                log.warn("Enrichment error ({}) for record: {}", enricherName, record, c.getLastFailure());

            });

            return Failsafe.with(retryPolicy, circuitBreaker, timeout)
                    .onSuccess(r -> {
                        metrics.success();
                        log.debug("Enrichment result is: {}", r);
                    })
                    .onFailure(failure -> {
                        //this is final failure after all retries
                        metrics.error();
                        log.warn("Enrichment error ({}) for record: {}", enricherName, record, failure.getFailure());
                    }).get(callable::call);
        }
```

I successfully compiled the module.

Note: There's some fixes in the suggestion like the warn log in the onFailure listener which now receive the exception from the failure.

guirenard commented on nuxeo-ai-core/pom.xml:
```diff
@@ -10,10 +10,10 @@
-      <artifactId>proto</artifactId>
-    </dependency>
+      <dependency>
+          <groupId>com.google.protobuf</groupId>
```
bad indent

You probably want to have a look to https://hyland.atlassian.net/wiki/spaces/NuxEng/pages/2232485639/Code+Formatting

debjit-hyland committed:
Addressed PR review comments

debjit-hyland committed:
Correct indentation
